lock/unlock: ccip eth sender/receiver contract

makramkd opened the pull request:
## Motivation


## Solution

makramkd committed:
wip

makramkd committed:
fix incorrect condition

makramkd committed:
fix receiving side

makramkd committed:
some more comments

makramkd committed:
additional condition for fee token

makramkd committed:
transfer weth if native call fails

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,206 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
```
0.8.19

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,206 @@
+
+import {Client} from "./../libraries/Client.sol";
+import {CCIPReceiver} from "./CCIPReceiver.sol";
+import {IWrappedNative} from "../interfaces/IWrappedNative.sol";
```
should be moved up to the first section

makramkd replied:
Fixed and sorted alphabetically.

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,206 @@
+
+  /// @notice the gas limit for the message call on the destination chain, 500,000 should be plenty.
+  /// @dev This won't vary on L2's, callbacks are always provided so-called "L2 gas".
+  uint256 public constant MESSAGE_GAS_LIMIT = 500_000;
```
Feels like massive overkill, wrapping is very cheap. Probably worth it to reduce this closer to the actual value

RensR replied:
Probably the default 200k is enough, would save gas/complexity if we can use that

makramkd replied:
Sounds good, do you think the default will always be enough?

makramkd replied:
A full transmit used 223k https://sepolia.etherscan.io/tx/0x74a52897369e4f39b9e986fb5c74d9eed7d73e157aa19e1aa1007f4102e568cc can probably drop this to 200k but will test some more (especially on sad paths) to make sure we have enough for the ERC-20 transfers.

makramkd replied:
Removed!

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,206 @@
+  /// @dev This won't vary on L2's, callbacks are always provided so-called "L2 gas".
+  uint256 public constant MESSAGE_GAS_LIMIT = 500_000;
+
+  string public constant override typeAndVersion = "EtherSenderReceiver 1.0.0-dev";
```
nit: i always like this to be the first property after the errors

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,206 @@
+    _validateMessage(message);
+
+    // set the gas limit for the call on destination.
+    bytes memory extraArgs = Client._argsToBytes(Client.EVMExtraArgsV1(MESSAGE_GAS_LIMIT));
```
nit: inline

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,206 @@
+    _validateFeeToken(message, msg.value);
+
+    // set the gas limit for the call on destination.
+    bytes memory extraArgs = Client._argsToBytes(Client.EVMExtraArgsV1(MESSAGE_GAS_LIMIT));
```
nit: inline

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,206 @@
+    // deposit the ether into the weth contract to get the wrapped ether.
+    // approve the router to spend the wrapped ether.
+    i_weth.deposit{value: message.tokenAmounts[0].amount}();
+    i_weth.approve(getRouter(), message.tokenAmounts[0].amount);
```
Can we do an infinite approval in the constructor?

makramkd replied:
I'm not sure how that'd work if `feeToken` is also the WETH token, since we do a `safeIncreaseAllowance`?

RensR replied:
You shouldn't :p just skip the approve if it's weth in that flow

makramkd replied:
Done.

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,206 @@
+      // Note: its not possible to have any leftover tokens in this path because we transferFrom the exact fee that CCIP
+      // requires from the caller.
+      IERC20(message.feeToken).safeTransferFrom(msg.sender, address(this), fee);
+      IERC20(message.feeToken).safeIncreaseAllowance(getRouter(), fee);
```
afaik safeIncreaseAllowance has the same number of issues the not-safe version has. In this case, it would make sense to just approve the exact amount, and not increase it, right?

makramkd replied:
Maybe `safeApprove` if that resets the approval to zero and then sets it to the actual amount?

makramkd replied:
Ah what about the case where `feeToken` is actually WETH?

We'd need to increase the approval in that case I think, cuz we approve a bit above the exact amount for the tokens themselves?

RensR replied:
infinite weth, check if weth -> if weth no extra approval, if not weth, normal approval

makramkd replied:
Done.

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,206 @@
+    // receiver and destination EOA addresses must be correctly specified.
+    // abi.decode will revert if the bytes in receiver do not decode to an address.
+    address receiver = abi.decode(message.receiver, (address));
+    if (receiver == address(0)) {
```
This is already done by CCIP, don't have to redo it imo

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,206 @@
+    Client.EVM2AnyMessage memory message
+  ) external payable returns (bytes32) {
+    _validateMessage(message);
+    _validateFeeToken(message, msg.value);
```
no need to pass msg.value, can just get it in the func

makramkd replied:
Ah true, I was considering calling this from `getFee`, but don't need to do this there.

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,206 @@
+  }
+
+  function _validateFeeToken(Client.EVM2AnyMessage memory message, uint256 msgValue) private pure {
+    Client.EVMTokenAmount memory tokenAmount = message.tokenAmounts[0];
```
Since this only cares about the amount, we can reduce gas by `message.tokenAmounts[0].amount`

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,206 @@
+  /// @return messageId The CCIP message ID.
+  function ccipSend(
+    uint64 destinationChainSelector,
+    Client.EVM2AnyMessage memory message
```
Can be calldata? Probably everywhere

makramkd replied:
Switched to calldata where possible, _ccipReceive cannot be calldata otherwise it doesn't correctly override _ccipReceive virtual in the CCIPReceiver base contract.

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,206 @@
+    // to the above _validateMessage and _validateFeeToken functions.
+
+    // decode the EOA receiver from the message data.
+    address eoaReceiver = abi.decode(message.data, (address));
```
doesn't have to be an eoa

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,206 @@
+    address eoaReceiver = abi.decode(message.data, (address));
+
+    // withdraw the WETH received from the token pool.
+    i_weth.withdraw(message.destTokenAmounts[0].amount);
```
message.destTokenAmounts[0].amount can be assigned to a var

makramkd committed:
address some PR comments

makramkd committed:
always fetch weth from router

to handle edge cases where weth address changes on the router for
whatever reason, this contract will still be functioning as expected.

makramkd committed:
switch some params to calldata, remove extra args setting

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -72,9 +71,7 @@ contract EtherSenderReceiver is CCIPReceiver, ITypeAndVersion {
-    // set the gas limit for the call on destination.
-    bytes memory extraArgs = Client._argsToBytes(Client.EVMExtraArgsV1(MESSAGE_GAS_LIMIT));
-    message.extraArgs = extraArgs;
+    message.extraArgs = Client._argsToBytes(Client.EVMExtraArgsV1(MESSAGE_GAS_LIMIT));
```
imo just kill setting this, is a lot cheaper and defaults to 200k

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -101,11 +98,9 @@ contract EtherSenderReceiver is CCIPReceiver, ITypeAndVersion {
-    // set the gas limit for the call on destination.
-    bytes memory extraArgs = Client._argsToBytes(Client.EVMExtraArgsV1(MESSAGE_GAS_LIMIT));
-    message.extraArgs = extraArgs;
+    message.extraArgs = Client._argsToBytes(Client.EVMExtraArgsV1(MESSAGE_GAS_LIMIT));
```
same, just kill setting this

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -158,19 +149,19 @@ contract EtherSenderReceiver is CCIPReceiver, ITypeAndVersion {
-      if (msgValue <= tokenAmount.amount) {
-        revert InsufficientMsgValue(tokenAmount.amount, msgValue);
+      // If the fee token is native, the fee must be included in msg.value.
+      if (msg.value <= tokenAmount) {
```
nit: having foundry tests would have shown that sweet sweet gas reduction 😭

makramkd replied:
Added some!

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,208 @@
+    uint64 destinationChainSelector,
+    Client.EVM2AnyMessage calldata message
+  ) external payable returns (bytes32) {
+    IWrappedNative weth = IWrappedNative(CCIPRouter(getRouter()).getWrappedNative());
```
I would maybe lean towards saying we just deploy a new EtherSenderReceiver when we change the wrapper native in CCIP. It's so insanely rare for us to do, and this adds so much ugly complexity and gas cost

makramkd replied:
Sounds good, will revert these changes, I think the path to migrating from this contract is straightforward.

makramkd committed:
add some tests

makramkd committed:
add more tests

makramkd committed:
bump snapshot

makramkd committed:
more tests

makramkd committed:
update comment, add fuzz tests

makramkd committed:
more tests

makramkd committed:
update script

makramkd committed:
gen wrappers

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,211 @@
+    if (validatedMessage.feeToken != address(0)) {
+      // Note: its not possible to have any leftover tokens in this path because we transferFrom the exact fee that CCIP
+      // requires from the caller.
+      // Note: this will fail if the caller didn't approve at least `fee` tokens to this contract.
```
nit: no need to start comments with Note:

makramkd replied:
Yeah I think I need to clean up these comments.

RensR commented on contracts/src/v0.8/ccip/applications/EtherSenderReceiver.sol:
```diff
@@ -0,0 +1,211 @@
+    // Note: we gave an infinite approval to the router in the constructor.
+    i_weth.deposit{value: validatedMessage.tokenAmounts[0].amount}();
+
+    // if the fee token is not native, we need to transfer the fee to this contract and re-approve it to the router.
```
nit: comment is for the line below

RensR committed:
tiny golf (#661)

makramkd committed:
add more fuzz tests, simplify logic

makramkd committed:
remove gas limit check

allow callers to set any gas limit they please, fall back to manual
execution on destination if not enough gas.

makramkd committed:
fix script

makramkd committed:
handle incorrect token and amounts in _ccipReceive

makramkd committed:
use vm.assume in fuzz tests

makramkd committed:
send over address(this).balance as ccip fee

this way excess ether never stays in the contract.

makramkd committed:
bump snapshot and wrappers

makramkd committed:
Merge branch 'ccip-develop' into makram/ethsendrecv

makramkd committed:
forge fmt

makramkd committed:
v1.5.0-dev version bump

makramkd committed:
Merge branch 'ccip-develop' into makram/ethsendrecv
